game/world/objects/npc.cpp

```diff
@@ -1345,29 +1348,48 @@ bool Npc::implGoTo(uint64_t dt, float destDist) {
   if(go2.flag==GT_No)
     return false;
 
-  if(isInAir()) {
+  if(isInAir() || interactive()!=nullptr) {
     mvAlgo.tick(dt);
     return true;
     }
 
-  auto dpos = go2.target()-position();
+  auto  dpos      = go2.target()-position();
+  float destDistY = 200.f;
 
   if(go2.flag==GT_Flee) {
     // nop
     }
   else if(mvAlgo.isClose(go2.target(),destDist)) {
+    if(std::abs(dpos.y)>destDistY && recalculateWayPath()) {
+      mvAlgo.tick(dt);
+      return true;
+      }
     bool finished = true;
     if(go2.flag==GT_Way) {
+      auto wpoint = go2.wp;
       go2.wp = wayPath.pop();
       if(go2.wp!=nullptr) {
+        if(wpoint!=nullptr && wpoint->ladder==go2.wp->ladder)
+          go2.ladder = go2.wp->ladder;
```

Comment:
Can you please clarify how this works:
`wpoint->ladder` and `go2.ladder` are generally thing, so what is reason to keep `go2.ladder` around?

Reply:
If two consecutive waypoints in `Waypath` have the same non `nullptr` interactive we know a ladder has to be used. Ladder can have any distance to the waypoint and if a new waypoint is popped  the information about the previous one is lost. The idea is to save ladder in `go2` and try `setInteraction` in every iteration. This works  because the edge intersects with ladder bbox so npc gets close eventually.

Reply:
Do I understand correctly:
in sequence of waypoints `w[0], w[1], w[2], w[3]` , if `w[1]` and `w[2]` point to same ladder npc need to use it, when approaching `w[1]` ?

would it be easier to implement `wayPath.peek()` to peek into next point without dequeuing it from list, so point can be poped, only after `setInteraction` succeeded?